### PR TITLE
Add warning to frequency_plan_id field when selecting none

### LINKED ISSUE #2805: Warn users about empty `frequency_plan_id` in the Console
#### Summary

After https://github.com/TheThingsNetwork/lorawan-stack/pull/2785 we allow creating/updating gateways without `frequency_plan_id` in the Console. While leaving empty `frequency_plan_id` is allowed we should warn the users about consequences.

#### Why do we need this?

Make users aware of limitations when creating/updating gateways without `frequency_plan_id`

#### What is already there? What do you see now?

The `frequency_plan_id` field which is not required:

Screenshot 2020-06-28 at 20 32 50

#### What is missing? What do you want to see?

@KrishnaIyer what are general limitations when creating gateways without `frequency_plan_id` ?

#### How do you propose to implement this?

I think the easiest is to show a warning that includes field description and the limitations. Once a frequency plan is selected we can remove the warning and show a regular description.

#### How do you propose to test this?

Check the `frequency_plan_id` field w/ and w/o value set.

#### Can you do this yourself and submit a Pull Request?

yes

### PR DESCRIPTION
**Pull request by mjamescompton**
#### Summary

Closes #2805

Visuals are the same as https://github.com/TheThingsNetwork/lorawan-stack/pull/2992 just shows only after selected the field.

#### Changes

- Allow manually adding additional options to fetch select.
- Add `no-frequency-plan` value to frequency plan select
- Add warning message if `no-frequency-plan` is selected
- Remove `no-frequency-plan`  if selected before submitting form. 

Before:
before_4

After:
after_4

#### Testing

Select `Do not use a frequency plan` in frequency plan dropdown. Try other values to check they can still be submitted 

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/webui/console/components/gateway-data-form/index.js [hunk 1]
```diff
@@ -164,6 +176,10 @@ class GatewayDataForm extends React.Component {
     }
   }
 
+  isNoFrequencyPlan(value) {
+    return Boolean(value === 'no-frequency-plan')
```
Comment (bafonins) [inline pkg/webui/console/components/gateway-data-form/index.js]:
```suggestion
    return value === EMPTY_FREQ_PLAN
```

File: pkg/webui/console/containers/freq-plans-select/index.js [hunk 2]
```diff
@@ -29,10 +29,13 @@ const m = defineMessages({
   title: 'Frequency plan',
   warning: 'Frequency plans unavailable',
   description: 'The frequency plan used by the end device',
+  none: 'Do not use a frequency plan',
```
Comment (bafonins) [inline pkg/webui/console/containers/freq-plans-select/index.js]:
```suggestion
  none: 'Do not set a frequency plan',
```

File: pkg/webui/console/containers/freq-plans-select/index.js [hunk 3]
```diff
@@ -29,10 +29,13 @@ const m = defineMessages({
...
 })
 
 const formatOptions = plans => plans.map(plan => ({ value: plan.id, label: plan.name }))
 
+const additionalGsFrequencyPlans = [{ value: 'no-frequency-plan', label: m.none.defaultMessage }]
```
Comment (bafonins) [inline pkg/webui/console/containers/freq-plans-select/index.js]: ?
```suggestion
const additionalGsFrequencyPlans = [{ value: 'no-frequency-plan', label: m.none }]
```

File: pkg/webui/console/containers/fetch-select/index.js [hunk 4]
```diff
@@ -33,11 +33,12 @@ export default function({
   defaultTitle,
   optionsFormatter = formatOptions,
   defaultDescription,
+  additionalOptions,
```
Comment (bafonins) [inline pkg/webui/console/containers/fetch-select/index.js]:
```suggestion
  additionalOptions = [],
```

This breaks e2e tests as not every usage of `CreateFetchSelect` has this parameter defined.

**Review by bafonins — CHANGES_REQUESTED** [review]

**Review by kschiffer — COMMENTED** [review]
Nothing to add.

**Review by kschiffer — CHANGES_REQUESTED** [review]
Actually, there are two UX problems:
- The warning should be displayed as soon as the "No frequency plan" option is selected (and not only when the field blurs)
- The warning should already appear initially, when the field is set to "No frequency plan" in the general settings

**Comment by mjamescompton** [comment]
> The warning should already appear initially, when the field is set to "No frequency plan" in the general settings 

Just a note to myself to update the general settings form `onSubmit` to remove "No frequency plan" value so that its set as blank

**Comment by mjamescompton** [comment]
Changing frequency plan from value to no value

**Comment by mjamescompton** [comment]
@kschiffer  @bafonins Thought it was clearer to request review now before rebasing onto https://github.com/TheThingsNetwork/lorawan-stack/pull/3364. Let me know if you want me to do this first?

**Comment by bafonins** [comment]
@mjamescompton please rebase

File: pkg/webui/console/components/gateway-data-form/index.js [hunk 5]
```diff
@@ -141,10 +164,27 @@ class GatewayDataForm extends React.Component {
     this.setState({ shouldDisplayWarning: this.isNotValidDuration(value) })
   }
 
+  @bind
+  setFrequencyPlanTouched() {
+    const { setFieldTouched } = this.props.formRef.current
+    setFieldTouched('frequency_plan_id')
```
Comment (kschiffer) [inline pkg/webui/console/components/gateway-data-form/index.js]: Actually, I think we can mark select fields touched by default, when a value was selected:

```diff
diff --git a/pkg/webui/components/select/index.js b/pkg/webui/components/select/index.js
index 66822b752..b99fe50bb 100644
--- a/pkg/webui/components/select/index.js
+++ b/pkg/webui/components/select/index.js
@@ -105,7 +105,7 @@ class Select extends React.PureComponent {
       await this.setState({ value })
     }
 
-    onChange(value)
+    onChange(value, true)
   }
 
   @bind
```

**Review by kschiffer — CHANGES_REQUESTED** [review]
Works as expected. See my remaining comment.

File: pkg/webui/console/components/gateway-data-form/index.js [hunk 6]
```diff
@@ -164,6 +204,10 @@ class GatewayDataForm extends React.Component {
     }
   }
 
+  isNoFrequencyPlan(value) {
```
Comment (bafonins) [inline pkg/webui/console/components/gateway-data-form/index.js]:
```suggestion
  isEmptyFrequencyPlan(value) {
```

Comment (bafonins) [reply pkg/webui/console/components/gateway-data-form/index.js]: you can move it outside of the component class

File: pkg/webui/console/components/gateway-data-form/index.js [hunk 7]
```diff
@@ -133,6 +143,19 @@ class GatewayDataForm extends React.Component {
 
     this.state = {
       shouldDisplayWarning: this.isNotValidDuration(props.initialValues.schedule_anytime_delay),
+      showFrequencyPlanWarning: this.isNoFrequencyPlan(props.initialValues.freqPlan),
```
Comment (bafonins) [inline pkg/webui/console/components/gateway-data-form/index.js]:
```suggestion
      showFrequencyPlanWarning: this.isNoFrequencyPlan(props.initialValues.frequency_plan_id) && props.update,
```

File: pkg/webui/console/components/gateway-data-form/index.js [hunk 8]
```diff
@@ -133,6 +143,19 @@ class GatewayDataForm extends React.Component {
...
+    }
+  }
+
+  componentDidMount() {
+    const { update } = this.props
+    const { values } = this.props.formRef.current
+
+    if (this.isNoFrequencyPlan(values.frequency_plan_id) && update) {
```
Comment (bafonins) [inline pkg/webui/console/components/gateway-data-form/index.js]: You already have all values in `initialValues`. No need to access internal form values.

```suggestion
 	const { update, initialValues } = this.props
    if (this.isEmptyFrequencyPlan(initialValues.frequency_plan_id) && update) {
      this.setState({
        showFrequencyPlanWarning: true,
      })
    }
```

Do we actually need this? Isnt it enough to check for empty `frequency_plan_id` in `construtor`?

File: pkg/webui/console/views/gateway-add/index.js [hunk 9]
```diff
@@ -119,6 +119,7 @@ export default class GatewayAdd extends React.Component {
               onSubmit={this.handleSubmit}
               initialValues={initialValues}
               update={false}
+              formRef={React.createRef()}
```
Comment (bafonins) [inline pkg/webui/console/views/gateway-add/index.js]: What is the purpose of passing ref as a prop without actually using it in `<GatewayAdd />`?

**Review by bafonins — CHANGES_REQUESTED** [review]

**Comment by bafonins** [comment]
@mjamescompton What is the status here?

**Comment by bafonins** [comment]
@mjamescompton

**Review by bafonins — APPROVED** [review]

**Review by kschiffer — APPROVED** [review]

**Pull request merged**